### PR TITLE
Fix the way how stream event subscription is handled.

### PR DESCRIPTION
**Pull request by aam**

### CODE REVIEW

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 1]
```diff
@@ -175,17 +175,17 @@ class VMService {
   Future<Null> get done => _peer.done;
 
   // Events
-  Stream<ServiceEvent> get onDebugEvent => onEvent('Debug');
-  Stream<ServiceEvent> get onExtensionEvent => onEvent('Extension');
+  Future<Stream<ServiceEvent>> get onDebugEvent => onEvent('Debug');
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/vmservice.dart]: Use of `Future<Stream<T>>` should never be necessary?

Comment (aam) [reply packages/flutter_tools/lib/src/vmservice.dart]: Interesting, I didn't know that. Why? What if we are not able to provide `Stream` right away? (which is basically the case here)

Comment (mravn-google) [reply packages/flutter_tools/lib/src/vmservice.dart]: I may be wrong of course, but I actually think any use of `Future<Stream<T>>` should result in a linter warning. One should always be able to just return a `Stream<T>` immediately and then delay the first event on that stream.

Comment (aam) [reply packages/flutter_tools/lib/src/vmservice.dart]: Sure, but how do you represent the case where a method that returns `Stream` actually have to `await` for something else? In our case we want to say `(await vmService.onTimelineEvent).listen` to indicate that we don't want continue with rest of `runFromSource` until we get the stream, which happens after vm confirms it received the event subscription.

Comment (mravn-google) [reply packages/flutter_tools/lib/src/vmservice.dart]: I'd probably contend that the method does two things and should be split.

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 2]
```diff
@@ -175,17 +175,17 @@ class VMService {
...
+  Future<Stream<ServiceEvent>> get onIsolateEvent => onEvent('Isolate');
+  Future<Stream<ServiceEvent>> get onTimelineEvent => onEvent('Timeline');
   // TODO(johnmccutchan): Add FlutterView events.
 
   // Listen for a specific event name.
-  Stream<ServiceEvent> onEvent(String streamId) {
+  Future<Stream<ServiceEvent>> onEvent(String streamId) async {
     _streamListen(streamId);
-    return _getEventController(streamId).stream;
+    return new Future<Stream<ServiceEvent>>.value(_getEventController(streamId).stream);
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/vmservice.dart]: Shouldn't we `await` on the previous line and then just leave this line as it is?

Comment (aam) [reply packages/flutter_tools/lib/src/vmservice.dart]: You are right.

File: packages/flutter_tools/lib/src/vmservice.dart (hunk 1 above)
Comment (tvolkert) [reply packages/flutter_tools/lib/src/vmservice.dart]: When the method that returns a `Stream` has to await something else, the right solution is to use `async*` and `yield`.

Comment (mravn-google) [reply packages/flutter_tools/lib/src/vmservice.dart]: I don't think that is the issue here. We want to await an asynchronous stream setup operation before calling runInView. I think it should be done in its own method, separate from the Stream getters

Comment (tvolkert) [reply packages/flutter_tools/lib/src/vmservice.dart]: Just for cleanliness' sake, I sent https://github.com/flutter/flutter/pull/16711

**Pull request merged**